### ISSUE TITLE
Forall should behave consistently if an empty variable is used

### ISSUE THREAD

**Issue by Alexander-Schiendorfer** [bmxitalia/LTNtorch#4]
### Problem
I have code where sometimes a variable might be empty due to no instances of that class being in the batch during training, e.g., when doing:

```
x_AOB_2 = ltn.Variable("x_2", data[labels == 2])
```
Then, a forall operation on a variable with an empty tensor leads to NaN, as in the following example:
```
import torch
import ltn

# p = 1 means arithmetic mean, that should be good instead of a sum
Forall = ltn.Quantifier(ltn.fuzzy_ops.AggregPMeanError(p=1, stable=True), quantifier="f")

# the variable x runs over 1., 2., 5., 7. and gives value of the predicate 0.1, 0.2, 0.5, 0.7
x_values = torch.tensor([[1.], [2.], [5.], [7.]])

X_var = ltn.Variable("x", x_values)
P = ltn.Predicate(func=lambda x: x/10.)  # P(x) = x/10

# show P(X_var) results
print("P(X_var) results:", P(X_var))

X_var = ltn.Variable("x", torch.tensor([]))  # empty variable
result = Forall(X_var, P(X_var))
#result = Forall(X_var, P(X_var), cond_vars=[X_var], cond_fn=lambda x: x.value <= 0.4)
print("Forall result:", result)
````
results in:
```
Forall result: LTNObject(value=tensor(nan), free_vars=[])
```

However, if I use the line with `cond_vars` and `cond_fn`, the result becomes 1.0 even if there are no values over which to aggregate: 
```
result = Forall(X_var, P(X_var), cond_vars=[X_var], cond_fn=lambda x: x.value <= 0.4)
```
results in:
```
Forall result: LTNObject(value=tensor(1., dtype=torch.float64), free_vars=[])
```
The NaN  values in batch-wise training led to problems with more complex formulas and I had to write a safe wrapper around them.
### Solution suggestion
* Forall should behave equally regardless of whether the domain of the variable becomes empty by filtering or by the chosen grounding
* Forall could get a `robust` flag analogous to `strict` to ensure that empty domains lead to 1.0

**Comment by tommasocarraro** [comment]
Hello!

Thank you for your comment!

It is a lot of time I do not work on this repository, but I remember this should have been done on purpose.

LTN tries to maximize the overall satisfaction of the logical rules in your knowledge base.

If a rule is NaN or a variable is empty, this means there is nothing to optimize or maximize. 

This means that it can be considered as fully true.

In the loss, if you have, for example, one rule that is fully true, this means that, automatically, this rule will play no contribution on the loss function during training. 

All the focus of the model will be put on the other formulas.

This is exactly what you want when a variable is empty or NaN.

**Comment by tommasocarraro** [comment]
Let's leave this issue open as I can brainstorm with other as I am out of the game since one good year. I do not remember each of the implementations.

Thank you for your patience!

**Comment by Alexander-Schiendorfer** [comment]
Absolutely true, but I had a Forall over a Forall - and then Forall raises a ValueError because NaN is not within [0, 1] - maybe really a "robust" semantics flag could make sense?

No rush - I'm just evaluating LTNtorch for an industrial project - great work!

**Comment by tommasocarraro** [comment]
Hi Alexander,
Sorry for my very very late reply.

Could you kindly provide a small notebook example where this does happen?

So I can give it a check and see if I spot something!

Please, comment it out so I understand your logic.

Thank you,
Tommaso
